### PR TITLE
Make partitioning utils QDQ aware so it does not break up QDQ node units

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

If the EP handles QDQ node units, we need to make sure we do not split those into different partitions.

Update the partitioning utils to be QDQ aware. If there are node units we process the logical nodes they represent instead of individual nodes. This ensure we process all nodes in a QDQ node unit at the same time so that they are always in the same partition.

### Motivation and Context

Fix one of the issues in #19590

### CODE REVIEW

**Comment by HectorSVC** [comment]
> CreateSupportedPartitions(const GraphViewer& graph_viewer,

QNN EP use this one.

---
Refers to: onnxruntime/core/providers/partitioning_utils.cc:409 in 956412b. [](commit_id = 956412bef53d3c52eb870f0644b6447623816e43, deletion_comment = False)

**Comment by HectorSVC** [comment]
>       if (!Contains(node_outputs, input)) {

Is it guaranteed that the node order is topological sorted? Otherwise, I think it's safe to build up the node_outputs first before processing any inputs.

---
Refers to: onnxruntime/core/providers/partitioning_utils.cc:328 in 956412b. [](commit_id = 956412bef53d3c52eb870f0644b6447623816e43, deletion_comment = False)

**Comment by skottmckay** [comment]
>       if (!Contains(node_outputs, input)) {

It is currently as we process nodes in topological order when doing partitioning.

We could process all outputs first before inputs. Is there a benefit apart from not requiring things to be topologically sorted if we do that? 

---
In reply to: [1980232432](https://github.com/microsoft/onnxruntime/pull/19723#issuecomment-1980232432) [](http://example.com/codeflow?ancestors=1980232432)

---
Refers to: onnxruntime/core/providers/partitioning_utils.cc:328 in 956412b. [](commit_id = 956412bef53d3c52eb870f0644b6447623816e43, deletion_comment = False)

**Comment by HectorSVC** [comment]
>       if (!Contains(node_outputs, input)) {

It happened when I was trying to fix it in my PR which made the nodes not in topological order any more. That's why I was asking. I was wondering in case someone else do that again without the awareness the code here which requires nodes in topological order.

---
In reply to: [1980424111](https://github.com/microsoft/onnxruntime/pull/19723#issuecomment-1980424111) [](http://example.com/codeflow?ancestors=1980424111,1980232432)

---
Refers to: onnxruntime/core/providers/partitioning_utils.cc:328 in 956412b. [](commit_id = 956412bef53d3c52eb870f0644b6447623816e43, deletion_comment = False)

**Review by HectorSVC — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/partitioning_utils.cc [hunk 1]
```diff
@@ -159,48 +182,88 @@ std::vector<std::vector<const Node*>> CreateSupportedPartitionNodeGroups(
...
+                      if (is_node_supported) {
+                        supported_group_border.insert(&output);
+                      }
+
+                      auto& output_node_in_degree = in_degree[output.Index()];
+                      --output_node_in_degree;
+
+                      if (output_node_in_degree == 0) {
+                        nodes_to_process.push_back(&output);
+                      }
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/partitioning_utils.cc:243]: Good point. Went through a number of incarnations and the final one has a lot more duplicated code.

File: onnxruntime/test/providers/partitioning_utils_test.cc [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    Node& cast_node = builder.AddNode("Cast", {equal_output}, {output});
+    cast_node.AddAttribute("to",
+                           static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT));
+  };
+
+  CheckAllNodesProcessed(build_model);
+}
+
+// TopK produces 2 outputs, one of which is used in a QDQ node group (Q of values output)
+// and the other (indices output) is not. A downstream node consumuing the indices output has an edge from the target
```
Comment (edgchen1) [inline onnxruntime/test/providers/partitioning_utils_test.cc:138]:
```suggestion
// and the other (indices output) is not. A downstream node consuming the indices output has an edge from the target
```

File: onnxruntime/core/providers/partitioning_utils.cc [hunk 3]
```diff
@@ -173,34 +200,63 @@ std::vector<std::vector<const Node*>> CreateSupportedPartitionNodeGroups(
...
+                    [&](const Node::EdgeEnd& edge_end) {
+                      const Node& n = edge_end.GetNode();
+                      const NodeUnit& downstream_node_unit = *node_unit_map->at(&n);
+                      const Node& output = downstream_node_unit.GetNode();
+
+                      process_downstream_node(output);
+                    });
+    } else {
+      std::for_each(node.OutputNodesBegin(), node.OutputNodesEnd(),
+                    [&](const Node& output) { process_downstream_node(output); });
```
Comment (edgchen1) [inline onnxruntime/core/providers/partitioning_utils.cc:249]:
```suggestion
      std::for_each(node.OutputNodesBegin(), node.OutputNodesEnd(), process_downstream_node);
```

**Review by edgchen1 — DISMISSED** [review]

File: onnxruntime/core/providers/qnn/qnn_execution_provider.cc [hunk 4]
```diff
@@ -534,44 +536,39 @@ QNNExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_viewer
   size_t num_of_supported_nodes = 0;
 
   // Create partitions from supported nodes.
-  {
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/qnn/qnn_execution_provider.cc:537]: It was. Didn't seem to add much value (some variables are freed slightly earlier but nothing significant, and not avoiding any naming clashes) vs the cost of pushing all the code further to the left (which in general tends to result in people using shorter variable names or having to split lines sooner).

I should have suggested turning off whitespace diffs so it wasn't as noticeable.

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**